api/v1alpha2/backstage_types.go

```diff
@@ -0,0 +1,302 @@
+	Deployment *apiextensionsv1.JSON `json:"deployment,omitempty"`
+}
+
+type RuntimeConfig struct {
+	// Name of ConfigMap containing Backstage runtime objects configuration
+	BackstageConfigName string `json:"backstageConfig,omitempty"`
+	// Name of ConfigMap containing LocalDb (PostgreSQL) runtime objects configuration
+	LocalDbConfigName string `json:"localDbConfig,omitempty"`
+}
+
+type Database struct {
+	// Control the creation of a local PostgreSQL DB. Set to false if using for example an external Database for Backstage.
+	// +optional
+	//+kubebuilder:default=true
+	EnableLocalDb *bool `json:"enableLocalDb,omitempty"`
+
+	// Name of the secret for database authentication. Optional.
+	// For a local database deployment (EnableLocalDb=true), a secret will be auto generated if it does not exist.
+	// The secret shall include information used for the database access.
+	// An example for PostgreSQL DB access:
+	// "POSTGRES_PASSWORD": "rl4s3Fh4ng3M4"
+	// "POSTGRES_PORT": "5432"
+	// "POSTGRES_USER": "postgres"
+	// "POSTGRESQL_ADMIN_PASSWORD": "rl4s3Fh4ng3M4"
+	// "POSTGRES_HOST": "backstage-psql-bs1"  # For local database, set to "backstage-psql-<CR name>".
+	AuthSecretName string `json:"authSecretName,omitempty"`
+}
+
+type Application struct {
+	// References to existing app-configs ConfigMap objects, that will be mounted as files in the specified mount path.
+	// Each element can be a reference to any ConfigMap or Secret,
+	// and will be mounted inside the main application container under a specified mount directory.
+	// Additionally, each file will be passed as a `--config /mount/path/to/configmap/key` to the
+	// main container args in the order of the entries defined in the AppConfigs list.
+	// But bear in mind that for a single ConfigMap element containing several filenames,
+	// the order in which those files will be appended to the main container args cannot be guaranteed.
+	// So if you want to pass multiple app-config files, it is recommended to pass one ConfigMap per app-config file.
+	// +optional
+	AppConfig *AppConfig `json:"appConfig,omitempty"`
+
+	// Reference to an existing ConfigMap for Dynamic Plugins.
+	// A new one will be generated with the default config if not set.
+	// The ConfigMap object must have an existing key named: 'dynamic-plugins.yaml'.
+	// +optional
+	DynamicPluginsConfigMapName string `json:"dynamicPluginsConfigMapName,omitempty"`
+
+	// References to existing Config objects to use as extra config files.
+	// They will be mounted as files in the specified mount path.
+	// Each element can be a reference to any ConfigMap or Secret.
+	// +optional
+	ExtraFiles *ExtraFiles `json:"extraFiles,omitempty"`
+
+	// Extra environment variables
+	// +optional
+	ExtraEnvs *ExtraEnvs `json:"extraEnvs,omitempty"`
+
+	// Number of desired replicas to set in the Backstage Deployment.
+	// Defaults to 1.
+	// +optional
+	//+kubebuilder:default=1
+	Replicas *int32 `json:"replicas,omitempty"`
```

Comment:
Also, what would be the behavior if for example both `spec.application.replicas` and `spec.deployment.spec.replicas` are set in the CR?

Reply:
I'd agree with you in case if do not have GA obligations (indeed "alpha" API allows to do pretty much "whatever you want"). It is definitely simpler way for development. But, it is what it is, and we have to support older API for some time (6 mo).  
Anyway, even if we have kind of choice now, I'd consider it as a good exercise for the case when we declare the API as more stable but need to change it, so, I see no reason why not to change the API version if the API is changed :). 

> Also, what would be the behavior if for example both spec.application.replicas and spec.deployment.spec.replicas are set in the CR?

It is a good point, I think it should be fixed, so new API always wins.

Reply:
> I'd agree with you in case if do not have GA obligations (indeed "alpha" API allows to do pretty much "whatever you want"). It is definitely simpler way for development. But, it is what it is, and we have to support older API for some time (6 mo).

Well, I meant, to remove the deprecated fields from the new `v1alpha2` version (not `v1alpha1`, which I agree still needs to be supported). Otherwise, my point was to say that if we want to keep these deprecated fields in `v1alpha2`, then maybe `v1alpha2` is not needed at all, and the new fields could simply be added to `v1alpha1`.

Reply:
Not sure I understand how to support those fields if we remove them. Thoughts?

Reply:
As documented in https://sdk.operatorframework.io/docs/best-practices/best-practices/, I guess one approach could be to implement CRD conversion here, to automatically convert older `v1alpha1` CRs to this new `v1alpha2`.
But, even if it may be an interesting exercise, we should still be somewhat thoughtful about how we approach this. I still think that we could add the new fields you are suggesting here to the existing `v1alpha1`; IMO adding new fields to the CRD is a compatible change that should **not** require a new API version.
We could just document the deprecated ones accordingly.

Reply:
As you could read in the NOTE above: "v1alpha2 is backward compatible with v1alpha1 **w/o data conversion**"

Not sure what exactly you mean by "been thoughtful", my point is to make it supportable which mean the API version should unequivocally identify the state. So if "v1alpha1" may mean several different (even compatible) combinations of config is not appropriate for supportable (GA) application.